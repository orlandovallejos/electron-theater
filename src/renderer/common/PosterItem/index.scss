.wrapper {
  width: 230px;
  height: auto;
  overflow: hidden;
  padding: 16px 8px;

  .imgWrapper {
    display: flex;

    img {
      width: 200px;
      border-radius: 7px;
      margin: auto;
      box-shadow: 0 5px 15px 5px rgb(0 0 0 / 10%);
      transition: box-shadow .5s,transform .5s;
      border: 1px solid transparent;

      &:hover {
        box-shadow: 0 5px 45px 5px rgb(0 0 0 / 5%);
        transform: translate(0,-5px);
        transition: box-shadow .5s,transform .5s;
        border-color: #e9e9e9;
      }
    }
  }

  .titleWrapper {
    margin: 12px 20px 0 20px;

    .title {
      font-size: 12px;
      color: #f5f5f5;
    }
  }

  .titleExtraWrapper {
    margin: 4px 20px;
    display: flex;
    justify-content: space-between;

    .extraItem {
      align-items: center;
      display: flex;
      color: #747474;

      ion-icon {
        margin: 0 5px 0 0;
      }
      .text {
        font-size: 12px;
        margin-top: 5px;
      }
    }
  }
}
