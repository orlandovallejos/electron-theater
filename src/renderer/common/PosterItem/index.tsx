/*------------------------------------------------*/
/* LIBRARIES
/*------------------------------------------------*/
import React from 'react';
import Vibrant from 'node-vibrant';
/*------------------------------------------------*/
/* INTERNAL DEPENDENCIES
/*------------------------------------------------*/
// Styles
import PosterItemWrapper, { ImageWrapper } from './style';
// Helpers
import imageHelper from '../../../helpers/image';

const DEFAULT_COLOR = '#fff';

type Props = {
  url: string;
  title: string;
  voteAverage: number;
  releaseDate: string;
};
const PosterItem = (props: Props) => {
  const { url, title, voteAverage, releaseDate } = props;
  const [borderVibrant, setBorderVibrant] =
    React.useState<string>(DEFAULT_COLOR);
  const [borderDarkVibrant, setBorderDarkVibrant] =
    React.useState<string>(DEFAULT_COLOR);

  const setBorderColors = async (imageURL: string) => {
    const palette = await Vibrant.from(
      imageHelper.getPosterImage(imageURL)
    ).getPalette();

    const colorVibrant = palette.Vibrant?.getHex();
    setBorderVibrant(colorVibrant || DEFAULT_COLOR);
    const colorDarkVibrant = palette.DarkVibrant?.getHex();
    setBorderDarkVibrant(colorDarkVibrant || DEFAULT_COLOR);
  };

  React.useEffect(() => {
    if (url) {
      setBorderColors(url);
    }
  }, [url]);

  return (
    <PosterItemWrapper>
      <ImageWrapper vibrant={borderVibrant} darkVibrant={borderDarkVibrant}>
        <div className="item">
          {url && <img src={imageHelper.getPosterImage(url)} alt={title} />}
        </div>
      </ImageWrapper>
      <div className="title-wrapper">
        <span className="title">{title}</span>
      </div>
      <div className="title-extra-wrapper">
        <div className="extra-item">
          <ion-icon name="star" />
          <span className="text">{voteAverage}</span>
        </div>
        <div className="extra-item">
          <ion-icon name="calendar-clear" />
          <span className="text">{releaseDate.substring(0, 4)}</span>
        </div>
      </div>
    </PosterItemWrapper>
  );
};

/*------------------------------------------------*/
/* EXPORTS
/*------------------------------------------------*/
export default PosterItem;
