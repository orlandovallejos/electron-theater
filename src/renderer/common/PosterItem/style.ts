/*------------------------------------------------*/
/* LIBRARIES
/*------------------------------------------------*/
import styled from 'styled-components';
import { transparentize } from 'polished';
/*------------------------------------------------*/
/* INTERNAL DEPENDENCIES
/*------------------------------------------------*/
import theme from 'renderer/themes/default';

const PosterItemWrapper = styled.div`
  width: 230px;
  height: auto;
  overflow: hidden;
  padding: 16px 8px;

  .title-wrapper {
    margin: 12px 20px 0 20px;

    .title {
      font-size: 12px;
      color: ${theme.fonts.color.primary};
    }
  }

  .title-extra-wrapper {
    margin: 4px 20px;
    display: flex;
    justify-content: space-between;

    .extra-item {
      align-items: center;
      display: flex;
      color: #747474;

      ion-icon {
        margin: 0 5px 0 0;
      }
      .text {
        font-size: 12px;
        margin-top: 5px;
      }
    }
  }
`;
PosterItemWrapper.displayName = 'PosterItemWrapper';

interface TitleProps {
  readonly vibrant: string;
  readonly darkVibrant: string;
}
const ImageWrapper = styled.div<TitleProps>`
  display: flex;

  .item {
    width: 204px;
    height: 304px;
    border-radius: 7px;
    margin: auto;
    transition: box-shadow 0.5s, transform 0.5s;

    img {
      width: 200px;
      border-radius: 7px;
      height: 300px;
      margin: 2px;
    }

    &:hover {
      box-shadow: ${(props) => `
        16px 12px 10px -9px ${transparentize(0.5, props.darkVibrant)},
        -8px 12px 10px -10px ${transparentize(0.5, props.darkVibrant)},
        16px -8px 10px -10px ${transparentize(0.5, props.darkVibrant)},
        -16px -12px 10px -10px ${transparentize(0.5, props.vibrant)},
        -16px 8px 10px -10px ${transparentize(0.5, props.vibrant)},
        0px -12px 10px -10px ${transparentize(0.5, props.vibrant)};
      `};
      transform: translate(0, -5px);
      transition: box-shadow 0.5s, transform 0.5s;
      background: ${(props) => `
        linear-gradient(to right, ${props.vibrant}, ${props.darkVibrant});
      `};
    }
  }
`;
ImageWrapper.displayName = 'ImageWrapper';

/*------------------------------------------------*/
/* EXPORTS
/*------------------------------------------------*/
export { ImageWrapper };
export default PosterItemWrapper;
