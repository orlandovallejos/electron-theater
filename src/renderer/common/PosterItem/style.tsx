/*------------------------------------------------*/
/* LIBRARIES
/*------------------------------------------------*/
import styled from 'styled-components';
import { transparentize } from 'polished';

const Wrapper = styled.div`
  width: 230px;
  height: auto;
  overflow: hidden;
  padding: 16px 8px;

  .title-wrapper {
    margin: 12px 20px 0 20px;

    .title {
      font-size: 12px;
      color: #f5f5f5;
    }
  }

  .title-extra-wrapper {
    margin: 4px 20px;
    display: flex;
    justify-content: space-between;

    .extra-item {
      align-items: center;
      display: flex;
      color: #747474;

      ion-icon {
        margin: 0 5px 0 0;
      }
      .text {
        font-size: 12px;
        margin-top: 5px;
      }
    }
  }
`;
Wrapper.displayName = 'Wrapper';

interface TitleProps {
  readonly vibrant: string;
  readonly darkVibrant: string;
}
const ImageWrapper = styled.div<TitleProps>`
  display: flex;

  .item {
    width: 204px;
    height: 304px;
    border-radius: 7px;
    margin: auto;
    transition: box-shadow 0.5s, transform 0.5s;

    img {
      width: 200px;
      border-radius: 7px;
      height: 300px;
      margin: 2px;
      /* display: none; */
    }

    &:hover {
      box-shadow: ${(props) => `
        14px 12px 10px -9px ${transparentize(0.5, props.darkVibrant)},
        -8px 12px 10px -10px ${transparentize(0.5, props.darkVibrant)},
        16px -8px 10px -10px ${transparentize(0.5, props.darkVibrant)},
        -12px -12px 10px -10px ${transparentize(0.5, props.vibrant)},
        -12px 8px 10px -10px ${transparentize(0.5, props.vibrant)},
        0px -12px 10px -10px ${transparentize(0.5, props.vibrant)};
      `};
      transform: translate(0, -5px);
      transition: box-shadow 0.5s, transform 0.5s;
      background: ${(props) => `
        linear-gradient(to right, ${props.vibrant}, ${props.darkVibrant});
      `};
    }
  }
`;
ImageWrapper.displayName = 'ImageWrapper';

/*------------------------------------------------*/
/* EXPORTS
/*------------------------------------------------*/
export { ImageWrapper };
export default Wrapper;
