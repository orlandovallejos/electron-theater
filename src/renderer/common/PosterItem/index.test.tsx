/*------------------------------------------------*/
/* LIBRARIES
/*------------------------------------------------*/
import React from 'react';
import { act } from 'react-test-renderer';
import shallow from 'enzyme/shallow';
/*------------------------------------------------*/
/* INTERNAL DEPENDENCIES
/*------------------------------------------------*/
import PosterItem from './index';

jest.mock('node-vibrant', () => {
  return {
    from: jest.fn(() => ({
      getPalette: jest.fn().mockResolvedValue({
        Vibrant: {
          getHex: jest.fn(() => 'blue'),
        },
        DarkVibrant: {
          getHex: jest.fn(() => 'red'),
        },
      }),
    })),
  };
});

describe('<PosterItem />', () => {
  afterAll(() => {
    jest.resetAllMocks();
  });

  it('should render a poster', async () => {
    const wrapper = shallow(
      <PosterItem
        id="movie-01"
        url="www.site.com/img.jpg"
        title="My movie"
        voteAverage={7.8}
        releaseDate="2018-05-15"
      />
    );

    // This refresh the component and renders the last state:
    await act(async () => {});

    expect(wrapper.debug()).toMatchSnapshot();
  });
});
