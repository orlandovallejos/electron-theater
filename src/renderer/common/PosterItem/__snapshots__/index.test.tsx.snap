// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PosterItem /> should render a poster 1`] = `
"<PosterItemWrapper>
  <ImageWrapper vibrant=\\"#fff\\" darkVibrant=\\"#fff\\">
    <div className=\\"item\\">
      <img src=\\"w780www.site.com/img.jpg\\" alt=\\"My movie\\" />
    </div>
  </ImageWrapper>
  <div className=\\"title-wrapper\\">
    <span className=\\"title\\">
      My movie
    </span>
  </div>
  <div className=\\"title-extra-wrapper\\">
    <div className=\\"extra-item\\">
      <ion-icon name=\\"star\\" />
      <span className=\\"text\\">
        7.8
      </span>
    </div>
    <div className=\\"extra-item\\">
      <ion-icon name=\\"calendar-clear\\" />
      <span className=\\"text\\">
        2018
      </span>
    </div>
  </div>
</PosterItemWrapper>"
`;
